### ISSUE TITLE
Distinguish optimization algorithms that satisfy constraints at every iteration from the others

### ISSUE THREAD

**Issue by nbedou** [scipy/scipy#4916]
The scipy.optimize.minimize documentation does not show which optimization algorithms satisfy constraints at every iteration, and which ones may violate the constraints.

At present, users may assume the constraints are always satisfied, which is not the case.

There is such a distinction in Matlab documentation (http://fr.mathworks.com/help/optim/ug/writing-constraints.html#br9p_ry).

**Comment by argriffing** [comment]
I am inclined to call the optimization code buggy if it passes out-of-bounds values (values outside of the box constraints) to the objective function, but I'm less certain about more general constraints.  In any case it is good to have more complete documentation.

**Comment by rgommers** [comment]
@nbedou was there a specific case that triggered opening this issue?

**Comment by pv** [comment]
I've seen several people being surprised by this (eg on stackoverflow),
so it would probably be good to explicitly add information on this to
the docs.

**Comment by nbedou** [comment]
@rgommers In my case the objective function was an interpolated function obtained with scipy.interpolate.interp1d. The constraints were to remain inside the range of known data points, to avoid extrapolation. But the minimize function calls sometimes the objective outside of the bounds. It is not hard to fix (for example wrapping the objective function with a penalty function), but it is a bit annoying because there is nothing about in the documentation.

I do not think it is a bug, because some optimization algorithms might have good reasons to evaluate the objective function outside of the feasible set. But I think the documentation must contain a warning.

**Comment by aarchiba** [comment]
Particularly for functions that evaluate derivatives numerically, it can be difficult to only evaluate within bounds. But indeed, the docs should be clearer about this. For box constraints, it is also feasible to transform the problem to an unbounded one, but scipy has no support for this; it could be done with a trampoline object, but there would be some thinking required to have a comfortable interface.

**Comment by Andreea-G** [comment]
I agree this should to be fixed in the documentation, but not in the actual code. I had this issue before. In some cases, the success of the minimization itself may depend on how well you design an artificial function that blows up outside of your bounds or if your constraints are not obeyed (this actually happened to me). So this has to be left for the user to define. Also, for instance _minimize_cobyla has a parameter 'catol' ('Tolerance (absolute) for constraint violations'), and OptimizeResult has 'maxcv' ('The maximum constraint violation'). Maybe for optimize.minimize something should be added in the 'Notes' under 'Constrained minimization'?

**Comment by anntzer** [comment]
> Particularly for functions that evaluate derivatives numerically, it can be difficult to only evaluate within bounds.

I am not a numerical analyst, but why can't the derivatives be evaluated with an asymmetric formula (`(f(x)-f(x-h))/h` instead of `(f(x+h)-f(x))/h`) at the edges?

**Comment by ev-br** [comment]
> why can't the derivatives be evaluated with an asymmetric formula ((f(x)-f(x-h))/h instead of (f(x+h)-f(x))/h) at the edges?

Some minimizers do something long these lines (you need to switch between central and one-sided schemes of the same approximation order, but that's a detail),  some older ones do not. 
eg, least_squares does it via _numdiff.approx_derivative

**Comment by mdhaber** [comment]
@antonior92 You looked into the different optimization methods quite a bit for the SciPy paper. Did you know which always respect bounds/other constraints?

The [`Bounds` documentation](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.Bounds.html#scipy.optimize.Bounds) has `keep_feasible`. Presumably it works for `trust-constr`, but probably not for the rest?

**Comment by antonior92** [comment]
That is right @mdhaber we do implement that for ``trust-const`` and it does exactly what this issue ask for. 

This parameter, however, does not work for the other methods. And I would assume that they do not have these guarantees.

**Comment by mdhaber** [comment]
OK. The documentation for `Bounds` objects should also be more explicit that the `keep_feasible` option is only valid for certain methods, referring the reader to the method-specific documentation.
